src/lightmap.cpp

```diff
@@ -490,98 +673,160 @@ void map::build_seen_cache( const tripoint &origin, const int target_z )
 
-    int part;
-    if ( vehicle *veh = veh_at( origin, part ) ) {
-        // We're inside a vehicle. Do mirror calcs.
-        std::vector<int> mirrors = veh->all_parts_with_feature(VPFLAG_EXTENDS_VISION, true);
-        // Do all the sight checks first to prevent fake multiple reflection
-        // from happening due to mirrors becoming visible due to processing order.
-        // Cameras are also handled here, so that we only need to get through all veh parts once
-        int cam_control = -1;
-        for (std::vector<int>::iterator m_it = mirrors.begin(); m_it != mirrors.end(); /* noop */) {
-            const auto mirror_pos = veh->global_pos() + veh->parts[*m_it].precalc[0];
-            // We can utilize the current state of the seen cache to determine
-            // if the player can see the mirror from their position.
-            if( !veh->part_info( *m_it ).has_flag( "CAMERA" ) &&
-                seen_cache[mirror_pos.x][mirror_pos.y] < LIGHT_TRANSPARENCY_SOLID + 0.1 ) {
-                m_it = mirrors.erase(m_it);
-            } else if( !veh->part_info( *m_it ).has_flag( "CAMERA_CONTROL" ) ) {
-                ++m_it;
-            } else {
-                if( origin.x == mirror_pos.x && origin.y == mirror_pos.y && veh->camera_on ) {
-                    cam_control = *m_it;
-                }
-                m_it = mirrors.erase( m_it );
-            }
+        castLight<0, -1, -1, 0, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
+        castLight<-1, 0, 0, -1, sight_calc, sight_check>(
+            seen_cache, transparency_cache, origin.x, origin.y, 0 );
+    } else {
+        if( origin.z == target_z ) {
+            seen_cache[origin.x][origin.y] = LIGHT_TRANSPARENCY_CLEAR;
         }
 
-        for( size_t i = 0; i < mirrors.size(); i++ ) {
-            const int &mirror = mirrors[i];
-            bool is_camera = veh->part_info( mirror ).has_flag( "CAMERA" );
-            if( is_camera && cam_control < 0 ) {
-                continue; // Player not at camera control, so cameras don't work
-            }
+        // Cache the caches (pointers to them)
+        std::array<const float (*)[MAPSIZE*SEEX][MAPSIZE*SEEY], OVERMAP_LAYERS> transparency_caches;
+        std::array<float (*)[MAPSIZE*SEEX][MAPSIZE*SEEY], OVERMAP_LAYERS> seen_caches;
+        std::array<const bool (*)[MAPSIZE*SEEX][MAPSIZE*SEEY], OVERMAP_LAYERS> floor_caches;
+        for( int z = -OVERMAP_DEPTH; z <= OVERMAP_HEIGHT; z++ ) {
+            auto &cur_cache = get_cache( z );
+            transparency_caches[z + OVERMAP_DEPTH] = &cur_cache.transparency_cache;
+            seen_caches[z + OVERMAP_DEPTH] = &cur_cache.seen_cache;
+            floor_caches[z + OVERMAP_DEPTH] = &cur_cache.floor_cache;
+        }
 
-            const auto mirror_pos = veh->global_pos() + veh->parts[mirror].precalc[0];
+        // Down
+        cast_zlight<0, 1, 0, 1, 0, 0, -1, sight_calc, sight_check>(
```

Comment:
> All calls to cast_zlight have the same value for the template parameter xz and yz: 0. Is that intended? Is there actually any need for those parameters (extensions planed for the future)?

I added them before I knew what I needed, so they may actually be useless. I thought I might be able to use them to map the small cones directly above and directly below the player, but it doesn't really sound like it would work.

> Also, the // Up and // Down comments are strange

I got the comments wrong there, what matters is the `zz` template parameter. `// Up` should be above the first `cast_zlight` with `zz == 1`.